### ISSUE TITLE
Ejabberd unable to load ets_cache NIF on OpenBSD

### ISSUE THREAD

**Issue by paulfariello** [processone/ejabberd#2374]
> What version of ejabberd are you using?

18.03

> What operating system (version) are you using?

OpenBSD 6.1

> How did you install ejabberd (source, package, distribution)?

Source

> What did not work as expected? Are there error messages in the log? What
> was the unexpected behavior? What was the expected result?

```
2018-04-09 10:07:54.239 [info] <0.61.0>@ejabberd_config:start:73 Loading configuration from /etc/ejabberd/ejabberd.yml               
2018-04-09 10:07:55.124 [warning] <0.61.0>@ejabberd_pkix:opt_type:189 option 's2s_certfile' is deprecated, use 'certfiles' instead      
2018-04-09 10:07:57.344 [warning] <0.375.0>@ejabberd_pkix:check_ca:621 failed to read CA directory /etc/ssl/certs: no such file or directory; configuring 'ca_p
ath' or 'ca_file' options might help                                                                                                             
2018-04-09 10:07:57.368 [error] <0.376.0> Failed to load NIF /usr/local/lib/cache_tab-1.0.13/priv/lib/ets_cache: Failed to load NIF library /usr/local/lib/cach
e_tab-1.0.13/priv/lib/ets_cache: 'Cannot load specified object' (load_failed)                                           
2018-04-09 10:07:57.368 [error] emulator Error in process <0.376.0> on node ejabberd@localhost with exit value:         
{load_failed,[{erlang,nif_error,[load_failed],[]},{ets_cache,load_nif,0,[{file,"src/ets_cache.erl"},{line,307}]},{code_server,'-handle_on_load/5-fun-0-',1,[{fi
le,"code_server.erl"},{line,1347}]}]}                                                                                               
2018-04-09 10:07:57.371 [warning] <0.377.0> The on_load function for module ets_cache returned {load_failed,[{erlang,nif_error,[load_failed],[]},{ets_cache,loa
d_nif,0,[{...}|...]},{code_server,'-handle_on_load/5-fun-0-',1,[...]}]}                                                           
2018-04-09 10:07:57.371 [error] <0.375.0>@ets_cache:new CRASH REPORT Process <0.375.0> with 0 neighbours exited with reason: call to undefined function ets_cac
he:new(ca_cache, [{max_size,1000},{cache_missed,true},{life_time,3600000}]) in gen_server:init_it/6 line 352                             
2018-04-09 10:07:57.371 [error] <0.325.0>@ets_cache:new Supervisor ejabberd_sup had child ejabberd_pkix started with ejabberd_pkix:start_link() at undefined ex
it with reason call to undefined function ets_cache:new(ca_cache, [{max_size,1000},{cache_missed,true},{life_time,3600000}]) in context start_error
2018-04-09 10:07:57.378 [critical] <0.61.0>@ejabberd_app:start:64 Failed to start ejabberd application: {error,{shutdown,{failed_to_start_child,ejabberd_pkix,{
undef,[{ets_cache,new,[ca_cache,[{max_size,1000},{cache_missed,true},{life_time,3600000}]],[]},{ejabberd_pkix,init,1,[{file,"src/ejabberd_pkix.erl"},{line,214}
]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,328}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}}}}
```

I've check that `ets_cache.beam` and `ets_cache.so` are correctly installed.
I can run cache_tab app correctly.

**Comment by cromain** [comment]
i'll try to reproduce this.
Q: your install seems odd, as cache_tab is installed in /usr/local/lib instead of /usr/local/erlang/lib
any reason for this ? I guess you try to start ejabberd with a custom ejabberdctl script...
See how mainstream ejabberdctl script works, and check about ERL_LIBS environment variable

**Comment by paulfariello** [comment]
I haven't changed the way I compile and install ejabberd since 17.04
```
AUTOMAKE_VERSION=1.15 AUTOCONF_VERSION=2.69 ./autogen.sh
./configure --disable-iconv --disable-pam ERL=/usr/local/bin/erl19 ERLC=/usr/local/bin/erlc19 EPMD=/usr/local/bin/epmd19 --sysconfdir=/etc --enable-user=_ejabberd --localstatedir=/var
gmake
```
I've check and cache_tab on version 17.04 is also in /usr/local/lib.

If you want me to try something in particular, just let me know.

**Comment by cromain** [comment]
i can not reproduce cache_tab issue.
again, it looks really odd to have it directly installed in /usr/local
while erlang is in /usr/local/lib/erlang, i recommend to get ejabberd in /ur/local/lib/ejabberd using --prefix option to configure. you should get a cleaner install like this
$ ls /usr/local/lib/ejabberd/lib/ 
base64url-0.0.1   eimp-1.0.4        ezlib-1.0.4       fast_xml-1.1.29   goldrush-0.1.9    jose-1.8.4        p1_oauth2-0.6.2   stringprep-1.0.11
cache_tab-1.0.13  ejabberd-18.03.45 fast_tls-1.0.21   fast_yaml-1.0.13  jiffy-0.14.8      lager-3.4.2       p1_utils-1.0.11   xmpp-1.1.20

from here, any erlang lib would be either in /usr/local/lib/erlang/lib (for otp) or /usr/local/lib/ejabberd/lib (for ejabberd and deps)

**Comment by paulfariello** [comment]
When you say using --prefix option. Do you mean using `--prefix /usr/local/lib/ejabberd` on the configure step of ejabberd? Seems a bit strange since it will install all the binaries (even ejabberdctl) in `/usr/local/lib/ejabberd/sbin/`. Anyway I've tried that and I still have the same issue:

```
2018-04-23 13:45:09.370 [info] <0.31.0> Application p1_utils started on node ejabberd@localhost
2018-04-23 13:45:09.383 [info] <0.31.0> Application fast_yaml started on node ejabberd@localhost
2018-04-23 13:45:09.427 [info] <0.31.0> Application fast_tls started on node ejabberd@localhost
2018-04-23 13:45:09.474 [info] <0.31.0> Application fast_xml started on node ejabberd@localhost
2018-04-23 13:45:09.484 [info] <0.31.0> Application stringprep started on node ejabberd@localhost
2018-04-23 13:45:09.503 [info] <0.31.0> Application xmpp started on node ejabberd@localhost
2018-04-23 13:45:09.520 [info] <0.31.0> Application cache_tab started on node ejabberd@localhost
2018-04-23 13:45:09.543 [info] <0.31.0> Application eimp started on node ejabberd@localhost
2018-04-23 13:45:09.544 [info] <0.61.0>@ejabberd_config:start:73 Loading configuration from /etc/ejabberd/ejabberd.yml
2018-04-23 13:45:12.155 [warning] <0.61.0>@ejabberd_pkix:opt_type:189 option 's2s_certfile' is deprecated, use 'certfiles' instead
2018-04-23 13:45:16.510 [warning] <0.382.0>@ejabberd_pkix:check_ca:621 failed to read CA directory /etc/ssl/certs: no such file or directory; configuring 'ca_path' or 'ca_file' options might help
2018-04-23 13:45:16.525 [error] <0.383.0> Failed to load NIF /usr/local/lib/ejabberd/lib/cache_tab-1.0.13/priv/lib/ets_cache: Failed to load NIF library /usr/local/lib/ejabberd/lib/cache_tab-1.0.13/priv/lib/ets_cache: 'Cannot load specified object' (load_failed)
2018-04-23 13:45:16.525 [error] emulator Error in process <0.383.0> on node ejabberd@localhost with exit value:
{load_failed,[{erlang,nif_error,[load_failed],[]},{ets_cache,load_nif,0,[{file,"src/ets_cache.erl"},{line,307}]},{code_server,'-handle_on_load/5-fun-0-',1,[{file,"code_server.erl"},{line,1347}]}]}
2018-04-23 13:45:16.527 [warning] <0.384.0> The on_load function for module ets_cache returned {load_failed,[{erlang,nif_error,[load_failed],[]},{ets_cache,load_nif,0,[{...}|...]},{code_server,'-handle_on_load/5-fun-0-',1,[...]}]}
2018-04-23 13:45:16.528 [error] <0.382.0>@ets_cache:new CRASH REPORT Process <0.382.0> with 0 neighbours exited with reason: call to undefined function ets_cache:new(ca_cache, [{max_size,1000},{cache_missed,true},{life_time,3600000}]) in gen_server:init_it/6 line 352
2018-04-23 13:45:16.528 [error] <0.332.0>@ets_cache:new Supervisor ejabberd_sup had child ejabberd_pkix started with ejabberd_pkix:start_link() at undefined exit with reason call to undefined function ets_cache:new(ca_cache, [{max_size,1000},{cache_missed,true},{life_time,3600000}]) in context start_error
2018-04-23 13:45:16.536 [critical] <0.61.0>@ejabberd_app:start:64 Failed to start ejabberd application: {error,{shutdown,{failed_to_start_child,ejabberd_pkix,{undef,[{ets_cache,new,[ca_cache,[{max_size,1000},{cache_missed,true},{life_time,3600000}]],[]},{ejabberd_pkix,init,1,[{file,"src/ejabberd_pkix.erl"},{line,214}]
},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,328}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}}}}
```

Maybe I need to use the prefix option directly on cache_tab. But how can I do that ?

Just in case you missed it, my 17.04 install works fine even with cache_tab and other libs installed directly in /usr/local/lib.

**Comment by paulfariello** [comment]
I ktrace/kdump ejabberd and found this in the log:
```
beam.smp:/usr/local/lib/ejabberd/lib/cache_tab-1.0.13/priv/lib/ets_cache.so: undefined symbol '__sync_add_and_fetch_8
```
Seems more likely to be my real issue.

**Comment by paulfariello** [comment]
Ok I've dig a little into it.
TLDR: I switch from egcc to clang. It works

FTR:
cache_tab do use `__sync_add_and_fetch`. Which I can compile and use in a simple test.c on my machine. But when I compile cache_tab I've an undefined symbol `__sync_add_and_fetch_8` on the final shared object. I couldn't find how the `sync_add_and_fetch` turns to `__sync_add_and_fetch_8`.
